File: Python/sorting/quicksort.py
def partition(array, low, high):    
    i = low - 1
    pivot = array[high]
 
    for j in range(low, high):
        if array[j] <= pivot:
            i = i + 1
            array[i], array[j] = array[j], array[i]

    array[i + 1], array[high] = array[high], array[i + 1]
    return i + 1

def quick_sort(array, low, high):
    if len(array) == 1:
        return array

    if low < high:
        part_index = partition(array, low, high)
        quick_sort(array, low, part_index - 1)
        quick_sort(array, part_index + 1, high)

#array = [1,8,6,3,9,99,121,2,3]
#quick_sort(array, 0, 8)
#print(array)

File: java/Sorting/QuickSort.java
package sorting;

import java.io.*;
import java.util.stream.Stream;

public class QuickSort {

  private static void qSort(int a[], int left, int right) {
    if (left < right) {

      int p = partition(a, left, right);

      qSort(a, left, p - 1);
      qSort(a, p + 1, right);
    }
  }

  private static int partition(int[] a, int l, int r) {
    int pivot = l;

    while (l <= r) {
    while (l <= r && a[l] <= a[pivot])
      l++;
    while (a[r] > a[pivot])
      r--;
    if (l < r)
      swap(a, l, r);
  }
  swap(a, pivot, r);
  return r;
}

  private static swap(int a, int i, int j){
    int temp = a[i];
    a[i] = a[j];
    a[j] = temp;
  }

  public static void main(String[] args) {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int[] arr = Stream.of(br.readLine().replaceAll("\\s+", " ").split(" ")).mapToInt(Integer :: parseInt).toArray();
    qSort(arr, 0, arr.length - 1);
    System.out.println(Arrays.toString(arr));
  }
}
